BlueprintUI/Sources/Element/ComparableElement.swift

```diff
@@ -0,0 +1,86 @@
+//
+//  ComparableElement.swift
+//  BlueprintUI
+//
+//  Created by Kyle Van Essen on 7/3/21.
+//
+
+import Foundation
+import UIKit
+
+
+/// An `ComparableElement` is an element which can cache its
+/// measurements and layouts across layout passes in order to dramatically improve
+/// performance of repeated layouts.
+///
+/// Some element types within Blueprint already conform to `ComparableElement`.
+/// If your element can be easily compared for equality, consider making your
+/// it conform to `ComparableElement` for improved performance across layouts.
+public protocol ComparableElement: AnyComparableElement {
+
+    ///
+    /// Indicates if the element is equivalent to the other provided element.
+    /// Return true if the elements are the same, or return false if something about
+    /// the element changed, such as attributes that affect measurement or layout.
+    ///
+    /// Note that even if this method returns true, the `ViewDescription`
+    /// backing the element will still be re-applied to the on-screen view.
+    ///
+    /// A `ComparableElementContext` parameter is provided, in case
+    /// you need access to the `Environment`, etc, to perform your comparison.
+    /// This is useful, for example, if your `Element` contains builders that are used to
+    /// build the final element structure.
+    ///
+    /// If your element conforms to `Equatable`, this method is synthesized automatically.
+    ///
+    func isEquivalent(to other: Self, in context: ComparableElementContext) -> Bool
+}
+
+
+/// A context object pased to `isEquivalent(to:in:)` that contains
+/// contextual information in which the `Element` is being compared. This
+/// is useful, for example, if you need information from the `Environment`
+/// to perform the comparison.
```

Comment:
A more concrete example might be helpful here, like when you'd need an `Environment` to actually build a `Builder` in order to compare the result (right?).

Reply:
Sure, I'll add an example to the protocol's func definition itself.

Reply:
Alternatively, we might just wanna remove this and then add it back once we find a real need where it matters in production – after thinking about this use case more (opaque builders that return elements); if the elements returned there are `ComparableElement`, _their_ layout will still be cached, so the upstream element that contains both known _and_ unknown element values can still benefit from caching.